test/framework/multi_template.go

```diff
@@ -0,0 +1,109 @@
+	EtcdMachineLabel = "cluster.x-k8s.io/etcd-cluster"
+	// workerMachineLabel is to get worker vSphere machines from machine list.
+	workerMachineLabel = "cluster.x-k8s.io/deployment-name"
+)
+
+// VSphereMultiTemplateUbuntu127RequiredEnvVars is required for TestVSphereMultipleTemplatesUbuntu127.
+var VSphereMultiTemplateUbuntu127RequiredEnvVars = []string{
+	VSphereMultiTemplateUbuntu127,
+}
+
+// RequiredVsphereMultiTemplateUbuntu127EnvVars return required env vars for TestVSphereMultipleTemplatesUbuntu127.
+func RequiredVsphereMultiTemplateUbuntu127EnvVars() []string {
+	return VSphereMultiTemplateUbuntu127RequiredEnvVars
+}
+
+// CheckVsphereMultiTemplateUbuntu127EnvVars checks is required env vars are present.
+func CheckVsphereMultiTemplateUbuntu127EnvVars(t *testing.T) {
+	checkRequiredEnvVars(t, VSphereMultiTemplateUbuntu127RequiredEnvVars)
+}
+
+// VsphereMachineValidation should return an error if either an error is encountered during execution or the validation logically fails.
+// This validation function will be executed by ValidateVsphereMachine and ValidateWorkerNodeVsphereMachine
+// with a vSphere machine config and a corresponding vSphere machine.
+type VsphereMachineValidation func(machineConfig *v1alpha1.VSphereMachineConfig, machine vspherev1.VSphereMachine) (err error)
+
+// ValidateVsphereMachine deduces the control plane or etcd configuration to machine mapping
+// and for each configuration/machine pair executes the provided validation functions.
+func (e *ClusterE2ETest) ValidateVsphereMachine(selector string, machineConfig *v1alpha1.VSphereMachineConfig, validations ...VsphereMachineValidation) {
+	e.T.Log("Validating VSphere machine template")
+	ctx := context.Background()
+	machines, err := e.KubectlClient.GetVsphereMachine(ctx, e.Cluster().KubeconfigFile, selector)
+	if err != nil {
+		e.T.Fatal(err)
+	}
+
+	for _, machine := range machines {
+		for _, validation := range validations {
+			err = validation(machineConfig, machine)
+			if err != nil {
+				e.T.Errorf("VSphere machine %v is not valid: %v", machine.Name, err)
+			}
+		}
+	}
+	e.StopIfFailed()
+}
+
+// ValidateWorkerNodeVsphereMachine deduces the worker node group configuration to machine mapping
+// and for each configuration/machine pair executes the provided validation functions.
+func (e *ClusterE2ETest) ValidateWorkerNodeVsphereMachine(validations ...VsphereMachineValidation) {
+	e.T.Log("Validating VSphere worker machine template")
+	ctx := context.Background()
+	machines, err := e.KubectlClient.GetVsphereMachine(ctx, e.Cluster().KubeconfigFile, workerMachineLabel)
+	if err != nil {
+		e.T.Fatal(err)
+	}
+
+	for _, wng := range e.ClusterConfig.Cluster.Spec.WorkerNodeGroupConfigurations {
+		machineFound := false
+		machineReg := regexp.MustCompile(common.WorkerMachineTemplateBase(e.ClusterName, wng.Name) + `-[0-9]+-[a-z0-9]+`)
+		for _, machine := range machines {
+			if !machineReg.MatchString(machine.Name) {
```

Comment:
Are we only doing a set of validations for one machine of the machine template? And not all the machines associated with that machine template?

Reply:
yeah it only validates one machine from each group, cuz each group use the same template